### ISSUE TITLE
[tern-es6] Tern_for_vim produces no output when used with tern-es6

### ISSUE THREAD

**Issue by benmccormick** [marijnh/tern#612]
When using tern for vim with tern-es6, I've been unable to acheive any of the desired tern functionality.  `TernDef`, `TernRefs`, `TernRename` etc all print out `''` and then do nothing.  I also don't see any additional information in autocomplete using YouCompleteMe.  All of this functionality works fine when I stick with the default tern install.

The only line that I'm changing in my vimrc is adding this:

```
let g:tern#command=['/usr/local/bin/node', '/Users/ben/Code/projects/tern-es6/bin/tern']
```

I've verified that my file paths are correct for both node and tern, and specified fully qualified paths to avoid any confusion with PATH.  I'm unsure of further debug steps to take at this point.

Environment Info: 

I'm using MacVim 7.4 on OSX 10.10.5.  I'm running version 0.12.7 of Node.  The latest commit on the tern-es6 repo is `2f77c5bf953d7774fe017abc69f744b0f8c8b44b` and I've run npm install within there.  I can run the bin/tern binary from the command line and it echos a port, but I'm not sure of the best way to debug it beyond that.

Note: I was unsure of whether to raise this issue in this repo or the tern_for_vim, but since it had already come up in a comment here: https://github.com/marijnh/tern/issues/116#issuecomment-132095392, and the editor plugin works fine with other versions of tern, I'm opening it here.  Please feel free to ask me to migrate it if you feel the issue is on the vim end.

**Comment by marijnh** [comment]
One possible way to diagnose the problem is to go to your project directory and start Tern manually (maybe `killall node` before to kill existing instances) with `/path/to/tern/bin/tern --verbose`, then run a Tern command from vim, and see what the server outputs.

I'll bring out a new release later today. You can also try if upgrading normally helps.

**Comment by polpo** [comment]
I'm seeing a few examples of this. One when using the node plugin and lodash (which I imagine many people use), and another with an ugly array literal.

__Example 1 (using lodash):__

.tern-project:
```
{
  "plugins": {
    "node": {}
  }
}
```

index.js:
```
var _ = require('lodash');
```

Calling `:TernDef` on the `_`.

Output of tern --verbose:
```
Request: {
  "files": [],
  "query": {
    "end": {
      "line": 0,
      "ch": 4
    },
    "lineCharPositions": true,
    "type": "definition",
    "file": "index.js"
  }
}
/home/iscott/.vim/bundle/tern_for_vim/node_modules/tern/lib/infer.js:535
    var c0 = str.charCodeAt(0)
                ^
TypeError: Cannot read property 'charCodeAt' of undefined
    at isInteger (/home/iscott/.vim/bundle/tern_for_vim/node_modules/tern/lib/infer.js:535:17)
    at extend.hasProp (/home/iscott/.vim/bundle/tern_for_vim/node_modules/tern/lib/infer.js:569:11)
    at extend.getProp (/home/iscott/.vim/bundle/tern_for_vim/node_modules/tern/lib/infer.js:601:24)
    at walk.simple.ObjectExpression (/home/iscott/.vim/bundle/tern_for_vim/node_modules/tern/plugin/doc_comment.js:84:44)
    at c (/home/iscott/.vim/bundle/tern_for_vim/node_modules/tern/node_modules/acorn/dist/walk.js:39:16)
    at Object.skipThrough (/home/iscott/.vim/bundle/tern_for_vim/node_modules/tern/node_modules/acorn/dist/walk.js:168:3)
    at c (/home/iscott/.vim/bundle/tern_for_vim/node_modules/tern/node_modules/acorn/dist/walk.js:38:15)
    at Object.exports.searchVisitor.walk.make.Property (/home/iscott/.vim/bundle/tern_for_vim/node_modules/tern/lib/infer.js:1930:35)
    at c (/home/iscott/.vim/bundle/tern_for_vim/node_modules/tern/node_modules/acorn/dist/walk.js:38:15)
    at Object.base.ObjectExpression (/home/iscott/.vim/bundle/tern_for_vim/node_modules/tern/node_modules/acorn/dist/walk.js:302:5)
```

Stepping into that line, it looks like `prop.key` has type `Literal`, not `Identifier`, and its value is `'_'`.

__Example 2 (ugly, yet completely legal array literal):__

index.js:
```
var foo;
foo = [,,'foo', 'bar'];
```

Output of `tern --verbose`:
```
Request: {
  "files": [
    {
      "text": "var foo;\nfoo = [,,'foo', 'bar'];\n\n",
      "type": "full",
      "name": "index.js"
    }
  ],
  "query": {
    "end": {
      "line": 1,
      "ch": 1
    },
    "lineCharPositions": true,
    "type": "definition",
    "file": "#0"
  }
}
/home/iscott/.vim/bundle/tern_for_vim/node_modules/tern/lib/infer.js:1613
    var handler = inferExprVisitor[node.type];
                                       ^
TypeError: Cannot read property 'type' of null
    at infer (/home/iscott/.vim/bundle/tern_for_vim/node_modules/tern/lib/infer.js:1613:40)
    at arrayLiteralType (/home/iscott/.vim/bundle/tern_for_vim/node_modules/tern/lib/infer.js:1344:9)
    at /home/iscott/.vim/bundle/tern_for_vim/node_modules/tern/lib/infer.js:1366:14
    at /home/iscott/.vim/bundle/tern_for_vim/node_modules/tern/lib/infer.js:1351:15
    at infer (/home/iscott/.vim/bundle/tern_for_vim/node_modules/tern/lib/infer.js:1614:22)
    at /home/iscott/.vim/bundle/tern_for_vim/node_modules/tern/lib/infer.js:1475:15
    at /home/iscott/.vim/bundle/tern_for_vim/node_modules/tern/lib/infer.js:1351:15
    at infer (/home/iscott/.vim/bundle/tern_for_vim/node_modules/tern/lib/infer.js:1614:22)
    at Object.walk.make.Expression (/home/iscott/.vim/bundle/tern_for_vim/node_modules/tern/lib/infer.js:1623:7)
    at c (/home/iscott/.vim/bundle/tern_for_vim/node_modules/tern/node_modules/acorn/dist/walk.js:68:35)
```

@benmccormick do you see something similar?

I'm guessing I should I file a separate issue for these.

**Comment by marijnh** [comment]
Thanks! The backtraces made it trivial to find the problem. Fixed in 4c8beadf4 and eebb630e54d9. 

@benmccormick Do these also solve your issues?

**Comment by benmccormick** [comment]
@marinjh my problem looks to have been related to referencing tern in a
location other than in node_modules of the vim plugin.  With the updated
code everything works fine for me.
